### PR TITLE
Feat: Major CLI foundational improvements

### PR DESCRIPTION
**Pull request by blainekasten**
# Summary 

This PR is a major rehaul of the CLI but unlocks a bunch of exciting new things. To end users, the changes should be nearly 100% opaque. I'll break this description up into a few chunks: What this previous issues are fixed from this, What is new because of this, and what's next.

## What this previous issues are fixed from this

There were several thorns about Click that get fixed by switching to Cyclopts. 

One of them was the awkward token parsing for the top level-shared parameters like `--api-key`. Previously you had to do it like `together --api-key [command] [command-args]`, with this you can specify these at any part of the path.

Another was about the way click handled runtime. It operated each level as an independent stack so we had to wrap commands with decorators to do error handlers, and telemetry log handlers. In this new world, we can handle all of this at a single top level point, greatly simplifying.

## What is new because of this

A bunch of new things are happening and most of it is around giving us finer control of all things from logic to aesthetics. Being built on top of `rich` we can use the theme to give common design to our application from the help pages to our internal `print()` calls.

Additionally we get a big performance boost in this in a few ways. First, Cyclopts is async so we can use the Async client and parallelize more work. Second, Cyclopts supports lazy loading commands. This means if you run `together files list` the files list code is loaded, but other modules aren't! This should improve the start time and memory overhead of using the CLI.

## what's next

This PR is generally meant to switch the code over and give us a foundation to build upon. The foundation includes standing up some basic components/design library which we will build upon in the future. We will also be working to support more human experience improvements in the future around prompting parameters.

<details>
  <summary><h3>Testing Plan</h3></summary>
  
  - [ ] Integration and Unit Tests pass
  - [ ] Output of each command has been verified and evidence pasted below
  - [ ] Proper review from Team members

**Output Evidence**

<details><summary><code>together</code> / <code>together --help</code></summary>
image
</details>

_endpoints_
<details><summary><code>together endpoints</code> / <code>together endpoints --help</code></summary>
image
</details>

<details><summary><code>together endpoints availability-zones</code></summary>
image
</details>

<details><summary><code>together endpoints availability-zones --json</code></summary>
image
</details>

<details><summary><code>together endpoints create</code></summary>
image
</details>

<details><summary><code>together endpoints create --model Qwen/Qwen2.5-7B</code></summary>
image
</details>

<details><summary><code>together endpoints create --model Qwen/Qwen2.5-7B  --hardware 2x_nvidia_h100_80gb_sxm</code></summary>
image
</details>

<details><summary><code>together endpoints create --model Qwen/Qwen2.5-7B  --hardware 2x_nvidia_h100_80gb_sxm --json</code></summary>
image
</details>

<details><summary><code>together endpoints delete</code></summary>
image
</details>

<details><summary><code>together endpoints delete endpoint-26b0e2a7-e554-431d-849c-8dbb23b96506</code></summary>
image
</details>

<details><summary><code>together endpoints delete endpoint-26b0e2a7-e554-431d-849c-8dbb23b96506 --json</code></summary>
image
</details>

</details>

### CODE REVIEW

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 1]
```diff
@@ -355,16 +372,16 @@ def _generate_dockerfile(config: Config) -> str:
         pip = """COPY pyproject.toml .
 RUN --mount=type=cache,target=/root/.cache/uv \\
     uv pip install --system --compile-bytecode . && \\
-    (python -c "import sprocket" 2>/dev/null || (echo "sprocket not found in pyproject.toml, installing from pypi.together.ai..." && uv pip install --system --extra-index-url https://pypi.together.ai/ sprocket))
+    (python -c "import sprocket" 2>/dev/null || (console.print "sprocket not found in pyproject.toml, installing from pypi.together.ai..." && uv pip install --system --extra-index-url https://pypi.together.ai/ sprocket))
```
Comment (technillogue) [inline src/together/lib/cli/api/beta/jig/jig.py]: incorrect here and underneath

File: src/together/lib/cli/__init__.py [hunk 2]
```diff
@@ -117,13 +89,345 @@ def track_request(request: httpx.Request) -> None:
...
+                "command": parsed_command,
+                "arguments": explicit_args,
+                "is_beta_command": is_beta_command,
+                "error": sanitize_cli_error_message(str(e)),
+            },
+        )
+
+        if debug:
+            raise e
+        elif isinstance(e, CycloptsError):
```
Comment (zainhas) [inline src/together/lib/cli/__init__.py:264]: I think we're exiting 0 on API errors here, the APIError branch inside run_command re-raises, but the outer handler only does sys.exit(1) for UnknownCommandError and "not isinstance APIError" cases. For plain APIError we fall through to finally and return normally.

Can you add an explicit sys.exit(1) for that path?

File: src/together/lib/cli/api/endpoints/_utils.py [hunk 3]
```diff
@@ -6,16 +6,16 @@
...
 F = TypeVar("F", bound=Callable[..., Any])
 
 
 def handle_endpoint_api_errors(prefix: str) -> Callable[[F], F]:
-    """Decorator to handle endpoint-specific API errors. Must be used with handle_api_errors."""
-    prefix_styled = click.style(f"{prefix}: ", fg="blue")
+    """Decorator to handle endpoint-specific API errors."""
 
     def decorator(f: F) -> F:
         @wraps(f)
```
Comment (zainhas) [inline src/together/lib/cli/api/endpoints/_utils.py:21]: Wraps async funcs with a sync wrapper; f(*args) returns a coroutine unawaited, so try/except APIError never fires. All user-friendly "not-found / permission" messages on availability_zones, list, retrieve, stop, update are dead code.

Fix: make wrapper async and await f(...)

File: src/together/lib/cli/api/fine_tuning/create.py [hunk 4]
```diff
@@ -305,52 +213,40 @@ def create(
...
-            if arg_source == ParameterSource.DEFAULT:
-                training_args[arg] = default_values[arg]
-
-        if ctx.get_parameter_source("lora_alpha") == ParameterSource.DEFAULT:  # type: ignore[attr-defined]
+            raise ValueError(f"LoRA fine-tuning is not supported for the model `{model}`")
+        # Cyclopts has no Click-style ctx/ParameterSource; use CLI defaults as heuristic for "unset".
+        if lora_r == DEFAULT_LORA_R:
+            training_args["lora_r"] = model_limits.lora_training.max_rank
+        if learning_rate == DEFAULT_LEARNING_RATE:
+            training_args["learning_rate"] = DEFAULT_LEARNING_RATE
```
Comment (zainhas) [inline src/together/lib/cli/api/fine_tuning/create.py]: assigns the SFT default learning rate back. Should be 1e-3 (LoRA default)

File: src/together/lib/cli/api/fine_tuning/download.py [hunk 5]
```diff
@@ -91,29 +78,36 @@ def download(
...
-    if json:
+    if config.json:
         os.environ.setdefault("TOGETHER_DISABLE_TQDM", "true")
 
     try:
-        file_path, file_size = DownloadManager(client).download(
+        sync_client = Together(
+            api_key=config.client.api_key,
+            # base_url=config.client.base_url,
+            # timeout=config.client.timeout,
```
Comment (zainhas) [inline src/together/lib/cli/api/fine_tuning/download.py]: inentional? this won't inherit base_url/timeout for downloads

File: src/together/lib/cli/api/evals/create.py [hunk 6]
```diff
@@ -20,269 +23,56 @@
...
-    help="Input template for model B.",
-)
-@click.pass_context
-@handle_api_errors("Evals")
-@auto_track_command
-def create(
-    ctx: click.Context,
-    type: Literal["classify", "score", "compare"],
+async def create(
+    type_: Literal["classify", "score", "compare"],
```
Comment (zainhas) [inline src/together/lib/cli/api/evals/create.py]: does cyclopts perform a override here to make sure that `together evals create --type` still works since the variable name here is "type_"?

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 7]
```diff
@@ -1273,12 +1221,262 @@ def volumes_describe(jig: Jig, name: str) -> Any:
...
+    config: Annotated[CLIConfig, Parameter(parse=False)],
+    toml_config: TomlConfigParameter = None,
+) -> None:
+    """List all deployments."""
+    _run_jig_cmd(config, toml_config, list_deployments)
+
+
+def secrets_set_cli(
+    name: Annotated[str, Parameter(name="--name", help="Secret name")],
+    value: Annotated[str, Parameter(name="--value", help="Secret value")],
```
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/jig.py:1393]: I know that we sanitize for telemetry but can we double check this won't leak secrets b/c here we choose to not use stdin/env?

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 8]
```diff
@@ -182,15 +186,19 @@ class Config:
 
     def __post_init__(self) -> None:
         if err := validate(self, type(self)):
-            raise click.UsageError(f"Invalid {self._path}: {err}")
+            raise Exception(f"Invalid {self._path}: {err}")
```
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/jig.py]: do we wanna use JigError here similar to how we use it below?

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 9]
```diff
@@ -182,15 +186,19 @@ class Config:
...
 
     @classmethod
-    def find(cls, config_path: str | None = None, init: bool = False) -> Config:
+    def find(cls, config_path: str | None = None, init: bool = False) -> JigConfig:
         """Find specified config_path, pyproject.toml, or jig.toml"""
         if config_path:
             found_path = Path(config_path)
             if not found_path.exists():
-                raise click.UsageError(f"Configuration file not found: {config_path}")
+                raise Exception(f"Configuration file not found: {config_path}")
```
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/jig.py]: same as above re. JigError

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 10]
```diff
@@ -203,10 +211,12 @@ def find(cls, config_path: str | None = None, init: bool = False) -> Config:
 
         if init:
             return cls()
-        raise click.UsageError("No pyproject.toml or jig.toml found, use --config to specify a config path")
+        raise Exception("No pyproject.toml or jig.toml found, use --config to specify a config path")
```
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/jig.py]: JigError

File: src/together/lib/cli/api/fine_tuning/cancel.py [hunk 11]
```diff
@@ -1,56 +1,53 @@
...
-            + click.style(job.status, fg="yellow"),
-            file=sys.stderr if json else None,
+        console.print(
+            f"[red]Training is not currently cancellable.[/red]\n\nCurrent status is [yellow]{job.status}[/yellow]",
         )
-        return
+        sys.exit(1)
 
-    if not quiet:
+    if not quiet or not config.non_interactive:
```
Comment (zainhas) [inline src/together/lib/cli/api/fine_tuning/cancel.py]: should this be: if not (quiet or config.non_interactive)

as is --quiet by itself won't suppress the prompt

File: src/together/lib/cli/_track_cli.py
Comment (zainhas) [inline src/together/lib/cli/_track_cli.py:239]: might be a tag for later but these regex's miss patterns like sk-(OAI keys)…, tog_…, hf_…, JWTs, URLs with embedded creds.

File: src/together/lib/cli/api/endpoints/update.py [hunk 12]
```diff
@@ -1,89 +1,49 @@
...
+    min_replicas: Optional[int] = None,
+    max_replicas: Optional[int] = None,
+    inactive_timeout: Optional[int] = None,
+    *,
+    config: Annotated[CLIConfig, Parameter(parse=False)],
 ) -> None:
     """Update a dedicated inference endpoint's configuration."""
-    if not any([display_name, min_replicas, max_replicas, inactive_timeout]):
-        click.secho("Error: At least one update option must be specified", fg="red", err=True)
+    if not any([display_name, min_replicas, max_replicas, inactive_timeout is not None]):
```
Comment (zainhas) [inline src/together/lib/cli/api/endpoints/update.py]: probably a bug from before but caught here now - min_replicas and max_replicas are evaluated by truthiness and b/c 0 is falsy in Python you technically cannot set min/max replicas to 0 by itself

prolly need a min_replicas is not None, max_replicas is not None aswell

File: src/together/lib/cli/api/files/upload.py [hunk 13]
```diff
@@ -1,71 +1,57 @@
...
+async def upload(
+    file: Annotated[Path, Parameter(required=True, help="The file to upload")],
+    purpose: Annotated[Optional[FilePurpose], Parameter(help="The purpose of the file")] = "fine-tune",
+    check: Annotated[Optional[bool], Parameter(help="Whether to check the file")] = True,
+    *,
+    config: Annotated[CLIConfig, Parameter(parse=False)],
+) -> None:
+    """Upload file."""
+    if config.json:
         os.environ.setdefault("TOGETHER_DISABLE_TQDM", "true")
```
Comment (zainhas) [inline src/together/lib/cli/api/files/upload.py:29]: will this make sure that the tqdm progress bar and the spinner dont collide?

File: src/together/lib/cli/__init__.py [hunk 14]
```diff
@@ -117,13 +89,345 @@ def track_request(request: httpx.Request) -> None:
...
+storage_app.command(
+    (f"{_CLI}.beta.jig.jig:jig_volumes_describe"),
+    name="describe",
+    help="Retrieve metadata for a volume from the Together platform",
+)
+storage_app.command((f"{_CLI}.beta.jig.jig:jig_volumes_list"), name="list", help="List volumes for a JIG deployment")
+
+
+def main() -> None:
+    install_completion(app)
```
Comment (zainhas) [inline src/together/lib/cli/__init__.py:431]: TOGETHER_NO_AUTO_COMPLETION is undocumented, do we want to display on first install since it modifies zshrc/bashrc?

File: src/together/lib/cli/api/beta/jig/jig.py [hunk 15]
```diff
@@ -620,10 +642,10 @@ def push(self, tag: str = "latest") -> None:
         if subprocess.run(login_cmd, input=self.together.api_key, text=True).returncode != 0:
             raise JigError("Registry login failed")
 
-        echo(f"Pushing {image}")
+        console.print(f"Pushing {image}")
         if subprocess.run(["docker", "push", image]).returncode != 0:
```
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/jig.py:636]: not sure about this but will these subprocess.run's in the cyclopts async runner block the event loop?

File: src/together/lib/cli/api/beta/jig/_uploader.py
Comment (zainhas) [inline src/together/lib/cli/api/beta/jig/_uploader.py:166]: not CLI code persay but can you double check with DCI if this sync client call inside async def is intentional - there are more below

File: src/together/lib/cli/api/evals/list.py [hunk 16]
```diff
@@ -1,60 +1,83 @@
...
+    table = ListTable("Evals")
+    table.add_primary_column("Workflow ID", ratio=2)
+    table.add_column("Type")
+    table.add_column("Status")
+    table.add_column("Model")
+    table.add_column("Model A")
+    table.add_column("Model B")
+
+    for job in data:
+        model = deep_get(job.parameters, ["model_to_evaluate", "model"], "")
```
Comment (zainhas) [inline src/together/lib/cli/api/evals/list.py]: this is failing and as a result I see this empty column

CleanShot 2026-04-20 at 15 57 06@2x

File: src/together/lib/cli/api/fine_tuning/retrieve.py [hunk 17]
```diff
@@ -1,36 +1,195 @@
...
 
-    if json:
-        print_json(openapi_dumps(response).decode("utf-8"))
+    if response.status in COMPLETED_STATUSES:
+        _print_job_details(response, fine_tune_id)
         return
 
-    # remove events from response for cleaner output
-    response.events = None
+    progress_text = generate_progress_bar(response, datetime.now().astimezone(), use_rich=True)
```
Comment (zainhas) [inline src/together/lib/cli/api/fine_tuning/retrieve.py:191]: should we use utc here just to be in sync with all the other timestamps we report?

**Review by zainhas — CHANGES_REQUESTED** [review]

File: src/together/lib/cli/api/fine_tuning/download.py (hunk 5 above)
Comment (blainekasten) [reply src/together/lib/cli/api/fine_tuning/download.py]: Yeah. trying to limit scope. We don't have an async based download manager. I'll add a TODO and make a ticket so we can implement this

File: src/together/lib/cli/api/fine_tuning/retrieve.py (hunk 17 above)
Comment (blainekasten) [reply src/together/lib/cli/api/fine_tuning/retrieve.py:191]: I don't think we're doing UTC anywhere in the CLI. Where do you see these?

File: src/together/lib/cli/api/evals/list.py (hunk 16 above)
Comment (blainekasten) [reply src/together/lib/cli/api/evals/list.py]: I think I fixed this. our eval api is poorly type defined

File: src/together/lib/cli/api/beta/jig/_uploader.py
Comment (blainekasten) [reply src/together/lib/cli/api/beta/jig/_uploader.py:166]: Yeah, i'll have them switch it all over to use async together once land

File: src/together/lib/cli/api/beta/jig/jig.py (hunk 15 above)
Comment (blainekasten) [reply src/together/lib/cli/api/beta/jig/jig.py:636]: hmm i'm not sure either. I feel confident though the DCI team will own and improve this now that it's async. I'll make sure to communicate with them

File: src/together/lib/cli/__init__.py (hunk 14 above)
Comment (blainekasten) [reply src/together/lib/cli/__init__.py:431]: I kinda think we should just drop the env var. You would have to know about it to not do it.

I think I'll just make it default to off for agents + ci

File: src/together/lib/cli/api/files/upload.py (hunk 13 above)
Comment (blainekasten) [reply src/together/lib/cli/api/files/upload.py:29]: I don't think they will collide currently just do to the way the logic steps are. But I'm going to drop the TQDM thing all together in a follow up PR. It was bad for the Download/Upload manager to print things in general. It should have done the logic and yielded to the caller to render to the console if it wanted to

File: src/together/lib/cli/api/beta/jig/jig.py (hunk 7 above)
Comment (blainekasten) [reply src/together/lib/cli/api/beta/jig/jig.py:1393]: Wow, super good find. It was leaking this, it was also just bugged where it was not handling positional arguments as arguments. But it's now fixed, ty for that great call out

File: src/together/lib/cli/api/files/upload.py (hunk 13 above)
Comment (blainekasten) [reply src/together/lib/cli/api/files/upload.py:29]: ticket: https://linear.app/together-ai/issue/DX-458/drop-tqdm-printing-from-file-uploader

File: src/together/lib/cli/_track_cli.py
Comment (blainekasten) [reply src/together/lib/cli/_track_cli.py:239]: Made ticket: https://linear.app/together-ai/issue/DX-459/improve-error-code-sanitization-for-cli-telemetry

File: src/together/lib/cli/api/evals/create.py (hunk 6 above)
Comment (blainekasten) [reply src/together/lib/cli/api/evals/create.py]: It must! /shrug

I also updated it to explicitly set the paramter name as `--type` so we don't worry about the framework regressing in the future

File: src/together/lib/cli/api/fine_tuning/create.py (hunk 4 above)
Comment (blainekasten) [reply src/together/lib/cli/api/fine_tuning/create.py]: fixed

File: src/together/lib/cli/api/fine_tuning/retrieve.py (hunk 17 above)
Comment (zainhas) [reply src/together/lib/cli/api/fine_tuning/retrieve.py:191]: i think those must be for logs than and not user facing

**Review by zainhas — APPROVED** [review]
